pkg/minikube/download/download_test.go

```diff
@@ -0,0 +1,190 @@
+/*
+Copyright 2020 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package download
+
+import (
+	"fmt"
+	"io/fs"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/go-logr/logr"
+	"k8s.io/klog/v2"
+	"k8s.io/minikube/pkg/minikube/constants"
+)
+
+type mockLogger struct {
+	downloads int
+	t         *testing.T
+}
+
+func (ml *mockLogger) Enabled() bool {
+	return true
+}
+func (ml *mockLogger) Info(msg string, keysAndValues ...interface{}) {
+	fmt.Println("hi - ", msg)
+	if strings.Contains(msg, "Mock download") {
+		// Make "downloads" take longer to increase lock time.
+		dur, err := time.ParseDuration("1s")
+		if err != nil {
+			ml.t.Errorf("Could not parse 1 second duration - should never happen")
+		}
+		time.Sleep(dur)
+
+		ml.downloads++
+	}
+}
+func (ml *mockLogger) Error(err error, msg string, keysAndValues ...interface{}) {}
+func (ml *mockLogger) V(level int) logr.Logger {
+	return ml
+}
+func (ml *mockLogger) WithValues(keysAndValues ...interface{}) logr.Logger {
+	return ml
+}
+func (ml *mockLogger) WithName(name string) logr.Logger {
+	return ml
+}
+
+func TestBinaryDownloadPreventsMultipleDownload(t *testing.T) {
+	EnableMock(true)
+	defer EnableMock(false)
+	tlog := &mockLogger{downloads: 0, t: t}
+
+	klog.SetLogger(tlog)
```

Comment:
this seems to be setting it globally. and unit tests happen in parallel.

Reply:
either set it for an instance of klog or make this test not parallel. preferably the first case

Reply:
Fixed in a different way! First I removed the dependence on klog by allowing to insert a "mocked" download implementation. Much cleaner and gives us much more flexibility in the future. I also reorganized download_test to force its tests to run in serial. Since each test file is compiled as its own binary, it is safe to modify the global state of download.go. Note other tests all set the download mock to the default create file mock, so these are all safe as well.